### PR TITLE
add more strict checking

### PR DESCRIPTION
**Pull request by lunatikub**
add more strict checking

### CODE REVIEW

File: jsmn.c [hunk 1]
```diff
@@ -229,18 +332,41 @@ int jsmn_parse(jsmn_parser *parser, const char *js, size_t len,
...
+					return JSMN_ERROR_INVAL;
+				}
+#endif
 				r = jsmn_parse_string(parser, js, len, tokens, num_tokens);
 				if (r < 0) return r;
 				count++;
 				if (parser->toksuper != -1 && tokens != NULL)
 					tokens[parser->toksuper].size++;
+#ifdef JSMN_STRICT
+				if (tokens != NULL) {
```
Comment (chlunde) [inline jsmn.c]: `parser->toksuper` might be -1

**Comment by lunatikub** [comment]
Any news ?

**Comment by edsiper** [comment]
@lunatikub 

I am using your patch in our project here:

https://github.com/fluent/fluent-bit/tree/master/lib/jsmn

I've found that when running in strict mode and tokens is NULL (because I want only to validate the string and not to populate results), the validation fails, a simple JSON map input like this helps to reproduce the problem:

```
{"key": "some value"}
``` 

e.g: 

```
    ret = jsmn_parse(&parser, json, len, NULL, 0);
```

any hints are appreciated.

**Comment by lunatikub** [comment]
Hi @edsiper,

Thanks for highlighting this important point in my patch:

The strict mode needs the array of tokens to validate a JSON: in the case of a string token, its needs to know the type of the parent to deduce the type of the next token by the function 'jsmn_string_next_tok'.

For instance, if a string is being parsed and the parent is an 'array', the next type after a comma must be a 'value'.

With this point in mind, I will modify the patch so that passing a NULL 'tokens' array will not be accepted by the function 'jsmn_parse'.

It will look like:
`#ifdef JSMN_STRICT
assert(tokens != NULL);
#endif`

Warm regards,
Thomas,

**Comment by edsiper** [comment]
@lunatikub thanks for the detailed info, that makes a lot of sense, I will go ahead and deprecate the usage of jsmn_parse() in that mode. 

btw, have you consider to maintain a fork of this project since it low activity from a maintainer perspective ?

**Comment by pt300** [comment]
More changes ahead.

**Pull request closed without merging**